Abstract a lot of code, clean up and improvments

MrJacz opened the pull request:
This PR serves to abstract a lot of code, clean up and add some improvements
- [x] Rewrite command & event loading
- [x] Rewrite logger class

MrJacz committed:
Starting point

MrJacz committed:
A lot, Just a lot

Commit:
stuff

Commit:
Extenders GuildMember: Clean up

Commit:
Responses: Nuke reboot messages

Commit:
Message Event: Add check to see if author is bot if so return

Commit:
misakiReady Event: Remove reboots stuff

Commit:
misakiReady Event: Add check for client status return if not ready

Commit:
guildMemberAdd Event: Remove useless check

Commit:
Reduce LOC

Commit:
Event store + More

Commit:
Await init

Commit:
general/announcement command: fix embed

Commit:
Commands: Update reminder & help, Social Command: update methods

Commit:
Abstract awaitReply

Commit:
Events: Fix ooops

Commit:
Make Extensions great again

Commit:
remove Embed.send because just no we dont need that

Commit:
Reduce LOC

YorkAARGH commented on base/Social.js:
```diff
@@ -39,32 +39,33 @@ class Social extends Command {
+
+(**y**es | **n**o)
+
+Reply with \`cancel\` to cancel the message. The message will timeout after 60 seconds.`, filter, 6000, null);
```
I dislike this being on multiple lines.

MrJacz replied:
same here

DevYukine replied:
if you dislike this i would recommend making awaitReply accepting arrays of strings aswell and give it an array of these

YorkAARGH commented on index.js:
```diff
@@ -1,176 +1,41 @@
+    "USER_NOTE_UPDATE",
+    "USER_SETTINGS_UPDATE",
+    "VOICE_SERVER_UPDATE",
+    "VOICE_STATE_UPDATE"],
```
I dislike how these are, pls 1 line them

MrJacz replied:
You cant read them when they're on one line.

DevYukine commented on base/Command.js:
```diff
@@ -84,5 +67,10 @@ class Command {
   }
+
+  async init() {
+    // ¯\_(ツ)_/¯
```
i would you please to remove this comment completly or make it actuall usefull information

Commit:
Reduce LOC, Clean up

YorkAARGH commented on events/message.js:
```diff
@@ -1,9 +1,29 @@
+    this.impliedPermissions = new Permissions([
+      "VIEW_CHANNEL",
+      "SEND_MESSAGES",
+      "SEND_TTS_MESSAGES",
```
It does not, and will not require `SEND_TTS_MESSAGES`

MrJacz replied:
It's implied permissions not what it requires.

YorkAARGH replied:
But still, it doesn't need that like at all

MrJacz replied:
I know it doesn't, But it is not permissions needed it is implied permissions..........

YorkAARGH commented on events/misakiReady.js:
```diff
@@ -3,7 +3,7 @@ const Event = require(`${process.cwd()}/base/Event.js`);
 
   async run() {
-    await this.client.methods.util.wait(500);
+    await this.client.methods.util.wait(750);
```
Why was this changed from `1000`? Seems like you're creeping back up to that.

MrJacz replied:
500ms was too short for command loading, 1 second is too long and 750 is just fine until i rewrite command and event loading

YorkAARGH commented on extenders/Message.js:
```diff
@@ -1,30 +1,28 @@
+    this.flags = [];
+  }
+
+  response(emoji = "❌", content, embed, options = {}) { // eslint-disable-line no-unused-vars
```
I'm pretty sure you can get rid of this.

MrJacz replied:
its used by way to many places

YorkAARGH replied:
So it is... I thought it was hardly used lol my bad

YorkAARGH commented on structures/MisakiClient.js:
```diff
@@ -68,17 +68,15 @@ class MisakiClient extends Client {
   async unloadCommand(commandPath, commandName) {
     const command = this.commands.get(commandName);
     if (!command) return `The command \`${commandName}\` doesn"t seem to exist, nor is it an alias. Try again!`;
-    if (command.shutdown) await command.shutdown();
```
Why was this removed? Isn't it needed for certain commands?

MrJacz replied:
No otherwise it would've been kept

YorkAARGH replied:
iirc, that was there if commands had an init function, for example York-Dev had init functions for the tags and examples

MrJacz replied:
Still not used in Misaki...

DevYukine commented on structures/MisakiClient.js:
```diff
@@ -0,0 +1,163 @@
+    this.config = require(`${process.cwd()}/config.js`);
+    this.logger = require(`${process.cwd()}/util/Logger`);
+    this.responses = require(`${process.cwd()}/assets/responses.js`);
+    this.idiotAPI = new idioticApi.Client(this.config.apiTokens.idiotToken, { dev: true });
```
i would recommend desctructuring here

YorkAARGH replied:
with how the api wrapper is written, de-structuring is basically impossible

DevYukine replied:
i dont say you should destructure the methods out of it, i just wanted that you descructure the Client out of the package, you can also rename `Client` to `IdioticClient` or something simular since the `Client` name is alrweady taken in that scope

DevYukine replied:
well actually you can ignore that its kinda pointless since you would need to rename it

DevYukine commented on util/Logger.js:
```diff
@@ -8,43 +8,31 @@ class Logger {
-      case "ready": {
-        return console.log(`${timestamp} ${chalk.black.bgGreen(type.toUpperCase())} ${content}`);
-      } 
+      case "log": return console.log(`${timestamp} ${chalk.bgBlue(type.toUpperCase())} ${content} `);
```
this is kinda redundant you maybe could solve this another way

MrJacz replied:
@Dev-Yukine What do you mean? I simply reduced unneeded LOC's

DevYukine replied:
rewrite the whole logic of this, all these cases are 95% similar and have the format "timestamp type_with_color content"

MrJacz replied:
well I was going to rewrite it anyway so yeah

DevYukine reviewed (dismissed):
im still not sure about the naming of "Command Store" and the rest but beside that lgtm still wait for @kyranet rewiev tho

AndrewJWin reviewed (changes requested):
There are too many changes in the help command, please take a look at the ``help.js`` file and review what changes are made in that, and amend your own.

DevYukine reviewed (commented):
After you rewritten the Logger class i will most likely approve

Commit:
Fix merge conflicts?

Commit:
Fix inconsistencies

Commit:
Make eslint great again

MrJacz committed:
Merge branch 'Yukine-Refactor' into Yukine-Refactor

Commit:
MisakiConsole is now a thing

Commit:
Merge branch 'Yukine-Refactor' of https://github.com/jacz-doesnt-like-forks-on-his-profile/Misaki into Yukine-Refactor

Commit:
new command loading, updated reload command, no need for the wait anymore

Commit:
fix commands

Commit:
Rewrite _flatten (#NoMoreKlasaCode)

Commit:
change awaitReply to array

Commit:
Reduce LOC

Commit:
Move arrDiff to util

Commit:
Move ratelimit from client to message event

Commit:
Move client.ratelimits to Message event

Commit:
Move levelCache to constructor

YorkAARGH commented on index.js:
```diff
@@ -1,173 +1,31 @@
+require(`${process.cwd()}/extenders/Guild.js`);
+require(`${process.cwd()}/extenders/DMChannel.js`);
+require(`${process.cwd()}/extenders/TextChannel.js`);
+const MisakiClient = require("./structures/MisakiClient.js");
```
This needs to conform to current practices.

MrJacz replied:
Im not sure why you use `${process.cwd()}` in require but its so not needed + breaks vs code intellisense

YorkAARGH replied:
When PR'ing in to an existing project, you should conform to what (minimal) standards they have, and this is one of the standards the project uses, pretty much every file uses `process.cwd()`, please change to conform.

MrJacz replied:
As agreed to on discord, I'll change the current style of using `${process.cwd}` back to the standard `./` and `../`

YorkAARGH commented on index.js:
```diff
@@ -1,173 +1,31 @@
-require("dotenv").load();
-require(`${process.cwd()}/extenders/Guild.js`);
+require("dotenv").config();
```
Why was this changed from `.load()` to `.config()`?

MrJacz replied:
Because `.load()` wasent loading the config. + The dotenv docs stated to use .config not .load nor can i see a .load method

YorkAARGH replied:
Some reason that's messed up, because it's working fine for Misaki in production.

MrJacz replied:
Odd

Commit:
Go back to standard require() format

Commit:
Change version to v2

This change has been approved by the developers.

Commit:
Add reload messages to reload command, Update reboot command a little, update reminder command a little too

MrJacz commented:
Requesting a review: @kyranet @Dev-Yukine @YorkAARGH 
I now feel this PR has done enough/served it purpose to rewrite/refactor/abstract/improve/clean the code, This has been a fun time, Doing something for a bot that doesn't use a framework is a little refreshing. I love writing my own things and not using framework methods, Misaki is a good example of how good a discord bot can be without a framework.

DevYukine reviewed (approved):
lgtm waiting for @kyranet tho

kyranet commented on base/Command.js:
```diff
@@ -1,52 +1,34 @@
+    this.usage = options.usage || "No usage provided.";
+    this.extended = options.extended || "No information provided.";
+    this.cost = options.cost || 0;
+    this.cooldown = options.cooldown || 5;
```
0 is falsy. If you wanted cooldown to be `0`, it'd become `5`.

```javascript
this.cooldown = 'cooldown' in options ? options.cooldown : 5;
```

kyranet commented on base/Event.js:
```diff
@@ -0,0 +1,41 @@
+    this.client = client;
+
+    this.name = options.name || file[file.length - 1].slice(0, -3);
+    this.enabled = options.enabled || true;
```
Same, `false` is falsy. If you set `EventOptions.enabled` to `false`, it'd become `true`.

kyranet commented on commands/animals/tiger.js:
```diff
@@ -19,11 +19,11 @@ class Tiger extends Social {
   async run(message, args, level, loadingMessage) {
     const { body } = await get("https://dashboard.typicalbot.com/api/v1/tiger").set("Authentication", process.env.TYPICAL);
-    await message.channel.buildEmbed()
+    const embed = this.client.methods.Embed()
```
You missed `new` here.

Commit:
Add urban from master + fix bug in urban

kyranet commented on commands/canvas/achievement.js:
```diff
@@ -17,18 +15,18 @@ class Achievement extends Social {
-    return Promise.resolve();
+  async cmdVerify(message, [...text], options) {
+    text = text.join(" ");
+    if (!text.length || text.length < 1) throw new this.client.methods.errors.UsageError("You must give an achievement description.", options.msg);
```
The second check is useless. Also `!text` always equals to `!text.length`.

Commit:
Cleanup

kyranet commented on commands/canvas/achievement.js:
```diff
@@ -17,18 +15,18 @@ class Achievement extends Social {
     if (message.mentions.users.first()) text = text.replace(/<@!?\d+>/, "").replace(/\n/g, " ").trim();
-    await message.channel.send(new MessageAttachment(await this.client.idiotAPI.achievement((message.mentions.users.first() || message.author).displayAvatarURL({ format:"png", size:32 }), text), "achievement.png"));
     await loadingMessage.delete();
+    return  message.channel.send(new this.client.methods.Attachment(await this.client.idiotAPI.achievement((message.mentions.users.first() || message.author).displayAvatarURL({ format:"png", size:32 }), text), "achievement.png"));
```
Remove the double space after `return`.

kyranet commented on commands/canvas/achievement.js:
```diff
@@ -17,18 +15,18 @@ class Achievement extends Social {
-    let text = args.join(" ");
+  async run(message, [...text], level, loadingMessage) {
+    text = text.join(" ");
     if (message.mentions.users.first()) text = text.replace(/<@!?\d+>/, "").replace(/\n/g, " ").trim();
```
```javascript
if (message.mentions.users.size) // ...
```

kyranet commented on commands/fun/say.js:
```diff
@@ -16,13 +15,13 @@ class Say extends Social {
     }
     const channel = message.guild.channels.get(channelid);
-    if (!message.member.permissionsIn(channel).has(["SEND_MESSAGES", "READ_MESSAGES"])) throw new UsageError("You do not have permission to `say` in that channel.", loadingMessage);
+    if (!message.member.permissionsIn(channel).has(["SEND_MESSAGES", "READ_MESSAGES"])) throw new this.client.methods.errors.UsageError("You do not have permission to `say` in that channel.", loadingMessage);
```
You can also put the array outside so it's shorter and easier to read and identify.

```javascript
if (channel.permissionsFor(message.member).missing(["SEND_MESSAGES", "READ_MESSAGES"]).length) throw new UsageError(/* ... */);
```

kyranet commented on commands/fun/urban.js:
```diff
@@ -0,0 +1,40 @@
+  }
+
+  async run(message, [...text]) { // eslint-disable-line no-unused-vars
+    if (!text.length) return message.send(message.getText("MUST_PROVIDE_SEARCH_TERM"));
```
`Message#send` is not a thing in Misaki, nor is `Message#getText`.

kyranet commented on commands/fun/urban.js:
```diff
@@ -0,0 +1,40 @@
+      .setDescription(description)
+      .setThumbnail("https://i.imgur.com/ressY86.png")
+      .setTitle(`${list[page].word} - Page ${page+1}/${list.length}`)
+      .addField(":thumbsup:", list[page].thumbs_up, true)
```
Could use unicode to improve readability 👍

kyranet commented on commands/fun/urban.js:
```diff
@@ -0,0 +1,40 @@
+      .addField(":thumbsup:", list[page].thumbs_up, true)
+      .addField(":thumbsdown:", list[page].thumbs_down, true)
+      .addBlankField(true)
+      .setURL(`${list[page].permalink}`)
```
Unnecesary template literal.

kyranet commented on commands/fun/xkcd.js:
```diff
@@ -24,13 +24,19 @@ class Xkdc extends Social {
   }
+
+  randomNum(min, max) {
+    min = Math.ceil(min);
```
This line is unnecesary, given that you always pass `1` to it.

kyranet commented on commands/fun/xkcd.js:
```diff
@@ -24,13 +24,19 @@ class Xkdc extends Social {
     }
     if (message.flags[0] == "r") {
-      const rn = await this.client.randomNum(1, ob.num);
+      const rn = await this.randomNum(1, ob.num);
```
Useless `await`.

kyranet commented on commands/reactions/blush.js:
```diff
@@ -15,8 +14,8 @@ class Blush extends Social {
     const target = message.mentions.members;
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to blush at.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot blush at yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to blush at.", loadingMessage));
```
Why is this sync method returning useless promises? Also, pick between `VAR === 0` and `!VAR`, but not both.

kyranet commented on commands/reactions/cry.js:
```diff
@@ -15,8 +14,8 @@ class Cry extends Social {
     const target = message.mentions.members;
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot cry on yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
```
Duplicated code in multiple files. Should be abstracted somewhere else taking both strings to reproduce the same effect.

kyranet commented on commands/social/games/slots.js:
```diff
@@ -36,12 +36,12 @@ class Slots extends Social {
     const results = machine.play();
     const winnings = this.help.cost * results.totalPoints;
-    message.buildEmbed()
+    const embed = this.client.methods.Embed()
```
You need `new ` here.

kyranet commented on commands/social/store/store.js:
```diff
@@ -125,7 +103,7 @@ class Store extends Social {
   
     const filter = m => m.author.id === message.author.id;
-    const response = await this.client.awaitReply(message, `Are you sure you want to purchase ${item.first().name} for ₲${item.first().price.toLocaleString()}?`, filter, undefined, null);
+    const response = await message.awaitReply(`Are you sure you want to purchase ${item.first().name} for ₲${item.first().price.toLocaleString()}?`, filter, undefined, null);
```
It should cache the first item.

kyranet commented on commands/system/reboot.js:
```diff
@@ -11,11 +11,8 @@ class Reboot extends Command {
-      await this.client.unloadCommand(cmd);
-    });
+  async run(message) {
+    await message.channel.send("Rebooting...");
```
Should catch if the message fails to send, so it runs the `process.exit`.

kyranet commented on commands/system/reload.js:
```diff
@@ -1,29 +1,36 @@
-    message.channel.send(`${this.client.responses.reloadSuccess.random().replaceAll("{{command}}", commands.help.name)}`);
+  resolvePiece(arg) {
+    const isCommand = this.client.commands.get(arg);
+    const isEvent = this.client.events.get(arg);
```
isEvent should be after the isCommand check.

kyranet commented on commands/utility/reminder.js:
```diff
@@ -14,13 +14,14 @@ class Reminder extends Command {
-    if (args.length === 0) {
+  async run(message, [...text]) {
+    text = text.join(" ");
+    if (!text.length) {
```
Pro tip: `text` is an empty array, therefore its length is 0, you can join it later when you need it as string.

kyranet commented on commands/utility/reminder.js:
```diff
@@ -14,13 +14,14 @@ class Reminder extends Command {
       else return message.channel.send("**Your Reminders:**\n" + reminders.map(r => `${r.reminder} - ${moment(r.reminderTimestamp).fromNow()}`).join("\n"));
     }
-    const blah = await this.regCheck(args.join(" "));
+    const blah = await this.regCheck(text);
```
`blah` is an indescribable name. I suggest using another name.

kyranet commented on commands/utility/reminder.js:
```diff
@@ -29,18 +30,16 @@ class Reminder extends Command {
   }
 
   regCheck(reminder) {
     const remind = /(?:^| )(?:in ?)?(((?:\d{1,2}(?:\.\d|\d)?)|a) ?((?:m(?:in(?:ute)?)?|h(?:our)?|d(?:ay)?|w(?:eek)?|m(?:onth)?|y(?:ear)?)s?))\b/gi.exec(reminder);
```
This supermassive RegExp could be outside this method.

kyranet commented on commands/utility/reminder.js:
```diff
@@ -29,18 +30,16 @@ class Reminder extends Command {
-    if (input.length === 0) return false;
+    const input = /(?:me ?)?(?:to ?)?(.*)/gi.exec(reminder)[1].replace(remind[0], "").trim();
+    if (!input.length) return false;
     return `${input}#${time}`;
```
Should return an array. What if the user decides to put a `#`? Exactly. It breaks.

kyranet commented on events/commandError.js:
```diff
@@ -1,19 +1,13 @@
-      }
+    if (error instanceof this.client.methods.errors.CustomError) {
+      if (error.msg) return error.msg.edit(error.message);
+      else return message.channel.send(error.message);
```
No else-return.

kyranet commented on events/guildMemberAdd.js:
```diff
@@ -14,6 +11,6 @@ module.exports = class {
     if (settings.welcomeEnabled !== "true") return;
     if (settings.welcomeType === "text") member.guild.channels.find("name", settings.welcomeChannel).send(`${this.client.emojis.get("396391329367588878")}  ${this.client.responses.welcomeMessages.random().replaceAll("{{user}}", member.user.username).replaceAll("{{amount}}", member.guild.memberCount).replaceAll("{{guild}}", member.guild.name).trim()}`).catch(console.error);
-    if (settings.welcomeType === "image") member.guild.channels.find("name", settings.welcomeChannel).send(new MessageAttachment(await this.client.idiotAPI.welcome("anime", member.user.bot, member.user.displayAvatarURL({ format: "png", size: 128 }), encodeURIComponent(member.user.tag), null, null))).catch(console.error);
+    if (settings.welcomeType === "image") member.guild.channels.find("name", settings.welcomeChannel).send(new this.client.methods.Attachment(await this.client.idiotAPI.welcome("anime", member.user.bot, member.user.displayAvatarURL({ format: "png", size: 128 }), encodeURIComponent(member.user.tag), null, null))).catch(console.error);
```
This line is superlong. Split it in different lines.

kyranet commented on events/message.js:
```diff
@@ -1,90 +1,111 @@
+    const cooldown = cmd.cooldown * 1000;
+    const ratelimits = this.ratelimits.get(message.author.id) || {}; // get the ENMAP first.
+    if (!ratelimits[cmd.name]) ratelimits[cmd.name] = Date.now() - cooldown; // see if the command has been run before if not, add the ratelimit
+    const differnce = Date.now() - ratelimits[cmd.name]; // easier to see the difference
```
difference*

kyranet commented on extenders/Guild.js:
```diff
@@ -1,15 +1,7 @@
-    }
-  };
+module.exports = Structures.extend("Guild", Guild => class extends Guild {
+  get store() {
```
Missing empty lines surrounding the getter.

kyranet commented on extenders/GuildMember.js:
```diff
@@ -1,58 +1,47 @@
+
+  constructor(...args) {
+    super(...args);
+    this.fullId = `${this.guild.id}-${this.id}`;
```
This should be a getter instead of contributing to a massive RAM usage.

kyranet commented on extenders/GuildMember.js:
```diff
@@ -1,58 +1,47 @@
+  }
+
+  giveItem(item, amount) {
+    this.inventory[item] += parseInt(amount);
```
Unnecesary `parseInt`. Same for the other methods.

kyranet commented on extenders/TextChannel.js:
```diff
@@ -1,15 +1,13 @@
-      
-    }
+  get readable() {
+    return !this.guild || this.permissionsFor(this.guild.me).has("VIEW_CHANNEL");
```
this.guild always exist in `TextChannel` instances.

kyranet commented on index.js:
```diff
@@ -1,173 +1,31 @@
-
-init();
+process.on("uncaughtException", err => {
+  const errorMsg = err.stack.replace(new RegExp(`${__dirname}/`, "g"), "./");
```
This RegExp should be cached.

kyranet commented on index.js:
```diff
@@ -1,173 +1,31 @@
-  .on("reconnect", () => client.logger.log("Bot reconnecting...", "log"))
-  .on("error", e => client.logger.error(e))
-  .on("warn", info => client.logger.warn(info));
+process.on("unhandledRejection", console.log);
```
Use `client.console.error` instead.

kyranet commented on commands/fun/xkcd.js:
```diff
@@ -24,13 +24,19 @@ class Xkdc extends Social {
       return message.channel.send({"embed": {"author":{"name":"Misaki | XKCD Comics"}, "description":`${ab.title} | #${ab.num}\n${ab.alt}`, "image":{"url":ab.img}}});
     }
     if (message.flags[0] == "r") {
```
eqeqeq.

kyranet commented on commands/reactions/bite.js:
```diff
@@ -15,8 +14,8 @@ class Bite extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to bite them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot bite yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to bite them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot bite yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/blush.js:
```diff
@@ -15,8 +14,8 @@ class Blush extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to blush at.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot blush at yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to blush at.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot blush at yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/cry.js:
```diff
@@ -15,8 +14,8 @@ class Cry extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot cry on yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot cry on yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/cuddle.js:
```diff
@@ -14,14 +13,14 @@ class Cuddle extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot cry on yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to cry on their shoulder.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot cry on yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/cuddle.js:
```diff
@@ -14,14 +13,14 @@ class Cuddle extends Social {
     const target = message.mentions.members;
-    if (target.size === 0) return message.response(undefined, "You need to mention someone to cuddle them.");
+    if (!target.size) return message.response(undefined, "You need to mention someone to cuddle them.");
     if (message.member == target.first()) return message.reponse(undefined, "You cannot cuddle yourself !");
```
eqeqeq.

kyranet commented on commands/reactions/hug.js:
```diff
@@ -15,8 +14,8 @@ class Hug extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to send them a hug.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot hug yourself !", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to send them a hug.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot hug yourself !", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/kiss.js:
```diff
@@ -15,8 +14,8 @@ class Kiss extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to kiss them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot kiss yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to kiss them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot kiss yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/lick.js:
```diff
@@ -15,8 +14,8 @@ class Lick extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to lick them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot lick yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to lick them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot lick yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/nom.js:
```diff
@@ -15,8 +14,8 @@ class Nom extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to nom on them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot nom yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to nom on them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot nom yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/pat.js:
```diff
@@ -15,8 +14,8 @@ class Pat extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to pat them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot pat yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to pat them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot pat yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/poke.js:
```diff
@@ -15,8 +14,8 @@ class Poke extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to poke them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot poke yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to poke them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot poke yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/punch.js:
```diff
@@ -15,8 +14,8 @@ class Punch extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to punch them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot punch yourself!", loadingMessage));
+    if (!target.size) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to punch them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot punch yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/slap.js:
```diff
@@ -15,8 +14,8 @@ class Slap extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to slap.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot slap yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to slap.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot slap yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/stare.js:
```diff
@@ -15,8 +14,8 @@ class Stare extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to stare at them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot stare at yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to stare at them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot stare at yourself!", loadingMessage));
```
eqeqeq.

kyranet commented on commands/reactions/tickle.js:
```diff
@@ -15,8 +14,8 @@ class Tickle extends Social {
-    if (target.size === 0) return Promise.reject(new UsageError("You need to mention someone to tickle them.", loadingMessage));
-    if (message.member == target.first()) return Promise.reject(new UsageError("You cannot tickle yourself!", loadingMessage));
+    if (target.size === 0) return Promise.reject(new this.client.methods.errors.UsageError("You need to mention someone to tickle them.", loadingMessage));
+    if (message.member == target.first()) return Promise.reject(new this.client.methods.errors.UsageError("You cannot tickle yourself!", loadingMessage));
```
eqeqeq.

kyranet reviewed (changes requested):
Many, many, requested changes (47). I also had to talk with @bdistin because CommandStore, EventStore and many other things (Command and Event aswell) are very similar to Klasa's old code with the RAM optimization PR patch applied. Nonetheless it's under a license that is not included.

Although his response was neutral, I'd feel better if at least some credits were given to Dirigeants for the original code.

There's also a lot of things that are strongly inspired by Klasa, for example, the `misakiReady` event (from `klasaReady`), our stores and pieces (without the Store and Piece abstract classes).

A lot of our style is applied in this PR, in case @YorkAARGH isn't aware, this code is way much better than the previous, yes, but it's so similar to Klasa, porting it would literally take hours as the way we handle the pieces, stores, loading, initialising, reloading, loading... are all identical.

Another thing that I highly dislike about this code is the amount of uncaught promises, there are many `message.channel.send` that aren't either awaited nor has a `.catch` callback. What does this mean? This leads to unhandled promise errors.

Sure, the `unhandledRejection` event handles those errors, but it's a bad practise: all errors should be caught from the code, not by the top layer from Node.js.

Another thing that I have noticed is that this code also lacks of documentation. Reducing lines of code isn't always a good thing, if it's undocumented, it's hard to read and understand what's going on. While the old code (and GuideBot) are highly documented, Misaki's seems to get less and less documentation. This leads to the project being harder for an external contributor to understand what is the code doing.

I must include that I have talked privately with you about MisakiConsole being ripped off from KlasaConsole (from Klasa 0.5.0) with the difference that colours are rather... hardcoded. But the design is very similar to our KlasaConsole rewrite made by bdistin. (#227 in dirigeants/klasa). I have, however, kindly asked for our license to be put in the top of the file. Instead, you renamed a method, removed the shard support, and also removed the notice about the code being inspired.

The MIT License allows you to use, modify and distribute our code, with the condition that our license must be put in the top of the file. Check our license [here](https://github.com/dirigeants/klasa/blob/master/LICENSE).

It's been fun to review all of this, I'm very familiar with this code so I understand it very well. Cheers and good luck!

MrJacz commented:
@kyranet 

> CommandStore, EventStore and many other things (Command and Event aswell) are very similar to Klasa's old code with the RAM optimization PR patch applied. Nonetheless it's under a license that is not included.

In fact, The EventStore was made by me. Originally it came from PenguBot where it had no loading stuff in the class but was the same (except it actually extended `Base` a class with some basic properties like client but nothing else) which I obviously removed that in this implementation but its really not a copy of Klasa's ways, Its inspired but it no way is it Klasa's code.

> There's also a lot of things that are strongly inspired by Klasa, for example, the misakiReady event (from klasaReady), our stores and pieces (without the Store and Piece abstract classes).

Funny thing is, I did this before I knew klasa was a thing

> A lot of our style is applied in this PR, in case @YorkAARGH isn't aware, this code is way much better than the previous, yes, but it's so similar to Klasa, porting it would literally take hours as the way we handle the pieces, stores, loading, initialising, reloading, loading... are all identical.

I'm quite sure York is aware but, The whole point of this PR was to make Misaki more understandable. That is very hard when you have A Client class in one file that handles the loading over commands & events with a directory scanning module (`klaw`) A couple of weeks ago I was whing about York using klaw you suggested the use of fs-nextra scan (Got to admit I had no clue on how to use this function) so me being me I looked at Klasa's store walk code (I knew it was used here) and used it. As I most likely would've messed up something if I didn't use a snippet so I did. But I really can't regret that as from the moment I say klaw in guidebot I disliked it so the use of a module we already have and use is put to use 🎉 

> Another thing that I highly dislike about this code is the amount of uncaught promises, there are many message.channel.send that aren't either awaited nor has a .catch callback. What does this mean? This leads to unhandled promise errors.

I can totally agree

> Another thing that I have noticed is that this code also lacks of documentation. Reducing lines of code isn't always a good thing, if it's undocumented, it's hard to read and understand what's going on. While the old code (and GuideBot) are highly documented, Misaki's seems to get less and less documentation. This leads to the project being harder for an external contributor to understand what is the code doing.

Yeah I can also agree on this, I'll work on adding documentation in the future + JSDoc

> I must include that I have talked privately with you about MisakiConsole being ripped off from KlasaConsole (from Klasa 0.5.0) with the difference that colours are rather... hardcoded. But the design is very similar to our KlasaConsole rewrite made by bdistin. (#227 in dirigeants/klasa). I have, however, kindly asked for our license to be put in the top of the file. Instead, you renamed a method, removed the shard support, and also removed the notice about the code being inspired.

Yes, we spoke, I don't see how this can really be a rip off honestly... Everything was off the top of my head except write, Now yes I had to look at snippets of code, no I didn't only look at klasa... I spent about 30 minutes reading a couple guides on the Console class & NodeJS stream's interesting, to say the least, but I really can't say this is a rip-off, I'd say write was inspired by KlasaConsole but that's it honestly

Commit:
Requested changes

Commit:
Achievement Command: Requested Changes

Commit:
Say Command: Requested changes

Commit:
Say Command: Requested changes

Commit:
xkcd Command: requested changes

Commit:
xkcd Command: requested changes

Commit:
reaction commands: requested changes

Commit:
Some more requested changes

Commit:
Requested changes

Commit:
Last requested change

Commit:
Make trump not restroy eslint again

bdistin commented:
>>There's also a lot of things that are strongly inspired by Klasa, for example, the misakiReady event (from klasaReady), our stores and pieces (without the Store and Piece abstract classes).

>Funny thing is, I did this before I knew klasa was a thing

Funny thing is, that almost all of the code used here in EventStore/Event (and more) is line for line (variable name for variable name even), identical code to a mixture of current and past implementations of klasa's version and by extension klasa's Piece/Stores. A coincidence you don't get without copying.

MrJacz commented:
@bdistin I've been looking at Event, Command pieces in klasa for the last 15 minutes, I can't find anything that even hints that them classes are a copy. Considering Command was a thing before I PRed that's a little hard to believe, the same variable names? So calling a variable name or file or client is copyright now? other then walkFiles the code is my own

Commit:
Rewrite Stores with my own spice

appellation commented:
I realize that I am not heavily involved in the development of Misaki, but york requested some code review so I will give it my best shot.

My main criticism of this project lies in ineffective application of objects and object prototypes.  While I will not take the time to make a fully-fledged, line-by-line review of this PR, I can generalize my findings and hopefully provide some insight on how to better utilize Javascript.  I realize that a lot of these issues existed prior to this PR, so don't take this as a criticism of only the OP.

1. **Unnecessary instance properties/methods:** while browsing the code in this PR, I noticed numerous classes that had instance properties that were constant or otherwise completely unrelated to the state of the class instance.  Class properties and methods should be reserved for things that either rely on or modify the instance state.  `client.methods` is a prime example of this, where every property is not reliant on the state of the client or any of it's properties.  Most people define methods on the client for convenience, which is what I imagine happened here.  However, it's is incredibly simple to just import the static util methods you require.  Oftentimes, these properties/methods should in fact be defined on another prototype, which lends the same utility in a more idiomatic and accessible fashion.  
This is by far my most significant criticism: throughout every command there seems to be at least one access of some client property that is actually static.  Besides being bad practice, you could see increased memory usage (from declaring multiple copies of the same static information) and bug occurrence (from inconsistent object mutation across instances).  My suggestion would be to enable the eslint rule `class-methods-use-this`; while it is not a rule that should be followed religiously, it will help you discover where you can cleanup some code and will cause some pause for thought before implementing new methods.

2. **Incorrect instance property/method location:** stuff like `message.settings` is immediately an example of such a problem.  A message cannot have settings, and so it's strange to see a property of them defined on the prototype.  Again, this frequently occurs in an effort for convenience, but is in fact confusing and could easily be defined on a more appropriate prototype.

Basically, my suggestion is as follows.  Don't be afraid to declare new classes as you need them; if you find yourself doing a lot of boilerplate to implement a feature, a class may be the thing for you.  For example, reminders might be something to consider writing a simple class for: perhaps you could consider defining `Member#reminders` as a store of reminder classes.  However, really think about why you're adding that particular instance property/method to that particular class and if it's not immediately related to what that class is about, perhaps you're better off finding another class (or making a new one).

Personally, I've found that defining a `Command` class (accepting a message in the constructor) with all your validation and command instance validation methods has been the most effective solution.  This also leads to your command execution being more object-oriented and thence friendlier.

When looking at a method signature, oftentimes the prototype that it *should* be defined on is accepted as one of the first few parameters.

On a personal note, I believe this PR achieves very little in the way of significant change.  Most of the work has been on the cleanup side of things, and calling this an abstraction is somewhat of a misnomer.

Other suggestions:

- The `console` dependency is entirely useless (and the internal console module is another example of point 1).  If you want decent logging, I would recommend switching to an async logging library like winston rather than one that just checks for console methods.
- I noticed that you're throwing custom error classes, which is great error handling practice.  However, there are some cases where you're using `if (!condition) return message.response(undefined, 'blah');`.  I'm unsure what the difference is, and I would recommend switching everything to `throw`.
  - Please fix that method so you don't have to send undefined as the first parameter.  It hurts.
- You should consider abstracting the argument system a little bit more.  I see that you have `awaitReply`, but numerous places duplicated code in the argument resolution.

MrJacz commented:
@appellation you're better to file a issue as half of this is not related to this PR what so ever

appellation commented:
This PR being titled "abstract code, cleanup, and improvements", I assumed that my comment was well within scope.
